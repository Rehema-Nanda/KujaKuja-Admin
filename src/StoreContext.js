// https://mobx-react.js.org/recipes-context

import React from "react";
import PropTypes from "prop-types";
import RootStore from "./stores/RootStore";

export const StoreContext = React.createContext(null);

export const StoreProvider = ({ children }) => {
    const store = new RootStore();
    return <StoreContext.Provider value={store}>{children}</StoreContext.Provider>;
};
StoreProvider.propTypes = {
    children: PropTypes.node.isRequired,
};

export const StoreConsumer = StoreContext.Consumer;

export const useStore = () => {
    const store = React.useContext(StoreContext);
    if (!store) {
        // this is especially useful in TypeScript so you don't need to be checking for null all the time
        throw new Error("useStore must be used within a StoreProvider.");
    }
    return store;
};
