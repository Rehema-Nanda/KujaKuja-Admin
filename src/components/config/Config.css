.title-text-input {
    width: 100%;
    padding: 5px;
    color: #333333;
    font-size: 15px;
    background-color: #EBEBEB !important;
}

input[type="file"] {
    padding-left: 300px;
    margin-right: -300px;
}
