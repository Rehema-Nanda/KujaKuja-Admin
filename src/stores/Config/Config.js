export default class Config {
    store = null;
    faviconUrl = null;
    highlightColour = null;
    logoUrl = null;
    titleText = null;
    logoFile = null
    faviconFile = null;

    constructor(store, json = null) {
        this.store = store;

        if (json) {
            this.updateFromJson(json);
        }
    }

    get asJson() {
        return {
            favicon_url: this.faviconUrl,
            highlight_colour: this.highlightColour,
            iso_two_letter_code: this.isoTwoLetterCode,
            logo_url: this.logoUrl,
            title_text: this.titleText,
            logo_file: this.logoFile,
            favicon_file: this.faviconFile,
        };
    }

    updateFromJson(json) {
        this.faviconUrl = json.favicon_url;
        this.highlightColour = json.highlight_colour;
        this.logoUrl = json.logo_url;
        this.titleText = json.title_text;
        this.logoFile = null;
        this.faviconFile = null;
    }

    async save() {
        const json = this.asJson;
        await this.store.apiService.putMultipathData(json);
    }
}
