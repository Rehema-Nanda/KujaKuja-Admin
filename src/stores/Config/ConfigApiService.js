import ApiService from "../ApiService";

export default class ConfigApiService extends ApiService {
    fetchConfig = async () => {
        const config = await this.makeHttpCall("config");
        return config.data.site_header;
    }

    putMultipathData = async (configJson) => {
        const data = new FormData();
        data.append("logo", configJson.logo_file);
        data.append("favicon", configJson.favicon_file);
        data.append("title_text", configJson.title_text);
        data.append("highlight_colour", configJson.highlight_colour);
        const config = {
            headers: {
                "Content-Type": "multipart/form-data",
                Accept: "application/json",
            },
            data: data,
        };
        const updateConfigResponse = await this.makeHttpCall("config/site_header", "put", config);
        return updateConfigResponse.data;
    }
}
