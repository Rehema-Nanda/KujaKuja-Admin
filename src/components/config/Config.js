import React from "react";
import { Col, Container, Row } from "reactstrap";

import "../Dashboard.css";
import "../../index.css";
import "./Config.css";
import reactCSS from "reactcss";
import { AvForm } from "availity-reactstrap-validation";
import { observer } from "mobx-react";
import { runInAction } from "mobx";
import { StoreContext } from "../../StoreContext";
import Banner from "../Banner";
import BtnRow from "../BtnRow";
import ColorPicker from "./ColorPicker";
import FileUploader from "./FileUploader";
import AppConfig from "../../AppConfig";

const hintHeader = require("../../img/config/hint-header-logo-220px.png");
const hintBrowser = require("../../img/config/hint-browser-tab-214px.png");
const hintColor = require("../../img/config/hint-color-band-213px.png");

class Config extends React.PureComponent {
    simpleState = {
        siteHeader: null,
    }

    constructor(props) {
        super(props);

        this.state = {
            refreshColorPicker: false,
            refreshHeaderLogo: false,
            refreshFaviconImage: false,
            buttonTitle: "Update",
            adviseUserToRefresh: false,
        };

        this.handleFormSubmit = this.handleFormSubmit.bind(this);
        this.handleHeaderHighlightColorChange = this.handleHeaderHighlightColorChange.bind(this);
        this.handleTitleTextChange = this.handleTitleTextChange.bind(this);
        this.updateHeaderFile = this.updateHeaderFile.bind(this);
        this.updateFaviconFile = this.updateFaviconFile.bind(this);
        this.cancelForm = this.cancelForm.bind(this);
    }

    async componentDidMount() {
        const { configStore } = this.context;
        const config = await configStore.loadConfig();
        await this.retrieveConfig(config);
    }


    async retrieveConfig(config) {
        const siteHeader = config;
        if (typeof siteHeader.titleText === "undefined" || siteHeader.titleText === null) {
            siteHeader.titleText = AppConfig.SITE_HEADER_DEFAULT_TITLE_TEXT;
        }
        if (typeof siteHeader.highlightColour === "undefined" || siteHeader.highlightColour === null) {
            siteHeader.highlightColour = AppConfig.SITE_HEADER_DEFAULT_HIGHLIGHT_COLOUR;
        }
        runInAction(() => {
            this.simpleState.siteHeader = siteHeader;
        });

        this.setState({
            adviseUserToRefresh: false,
        }, () => {
            this.refreshColorPicker();
            this.refreshHeaderLogo();
            this.refreshFaviconImage();
        });
    }

    refreshColorPicker() {
        const { refreshColorPicker } = this.state;

        this.setState({
            refreshColorPicker: !refreshColorPicker,
        });
    }

    refreshHeaderLogo() {
        const { refreshHeaderLogo } = this.state;

        this.setState({
            refreshHeaderLogo: !refreshHeaderLogo,
        });
    }

    refreshFaviconImage() {
        const { refreshFaviconImage } = this.state;

        this.setState({
            refreshFaviconImage: !refreshFaviconImage,
        });
    }

    updateHeaderFile(file) {
        const { refreshHeaderLogo } = this.state;

        this.simpleState.siteHeader.logoFile = file;

        this.setState({
            refreshHeaderLogo: !refreshHeaderLogo,
        });
    }

    updateFaviconFile(file) {
        const { refreshFaviconImage } = this.state;
        this.simpleState.siteHeader.faviconFile = file;
        this.setState({
            refreshFaviconImage: !refreshFaviconImage,
        });
    }

    handleHeaderHighlightColorChange(colour) {
        this.simpleState.siteHeader.highlightColour = colour.hex;
    }

    handleTitleTextChange(event) {
        this.simpleState.siteHeader.titleText = event.target.value;
    }

    async handleFormSubmit() {
        this.setState({
            buttonTitle: "Updating...",
            adviseUserToRefresh: false,
        }, () => {
            this.simpleState.siteHeader.save();
            this.setState({
                buttonTitle: "âœ” Updated!",
                adviseUserToRefresh: true,
            }, () => {
                setTimeout(() => {
                    this.setState({
                        buttonTitle: "Update",
                    });
                }, 3000);
            });
        });
    }

    async cancelForm() {
        await this.retrieveConfig();
    }

    render() {
        const {
            adviseUserToRefresh, refreshColorPicker, refreshHeaderLogo, refreshFaviconImage, buttonTitle,
        } = this.state;
        const styles = reactCSS({
            default: {
                configParent: {
                    backgroundColor: "#FFFFFF",
                    borderRadius: "8px",
                    margin: "10px auto 56px auto",
                    boxShadow: "0 24px 18px -10px rgba(23,29,33,0.08)",
                    width: "100%",
                    maxWidth: "800px",
                    padding: "10px",
                },
                hintImage: {
                    width: "200px",
                    marginTop: "-10px",
                    marginLeft: "-10px",
                    marginRight: "10px",
                },
                settingsParent: {
                    marginTop: "30px",
                },
                explainerHeader: {
                    textAlign: "left",
                    fontSize: "18px",
                },
                RowParent: {
                    backgroundColor: "#F5F5F5",
                    borderRadius: "8px",
                    margin: "10px",
                    padding: "20px 10px",
                },
                RowParent2: {
                    marginTop: "20px",
                },
                titleTextParent: {
                    backgroundColor: "#FFFFFF",
                    padding: "10px",
                    borderRadius: "5px",
                    minWidth: "300px",
                    position: "relative",
                },
                title: {
                    color: "#5b5656",
                    fontSize: "15px",
                    textAlign: "left",
                    marginBottom: "0",
                },
                titleTextHeader: {
                    marginTop: "-8px",
                },
                buttonParent: {
                    textAlign: "center",
                    marginTop: "20px",
                },
                hint: {
                    textAlign: "center",
                    padding: "0",
                },
                hintVisibile: {
                    display: adviseUserToRefresh ? "block" : "none",
                },
            },
        });
        const { siteHeader } = this.simpleState;

        return (

            <div className="page">

                <div className="title-background-parent">
                    <div className="title-background-shape" />
                </div>

                <Banner pageTitle="custom header" />
                <Container>

                    <AvForm
                        encType="multipart/form-data"
                        onValidSubmit={this.handleFormSubmit}
                        model={siteHeader}
                    >

                        <div style={styles.configParent}>
                            <Row style={styles.RowParent}>

                                <Col lg={{ size: 12 }}>
                                    <h4 style={styles.explainerHeader}>Header Highlight Color</h4>
                                    <p>This is the thin color band that runs across the very top of the page.</p>

                                    <div style={styles.settingsParent} className="d-flex justify-content-start">
                                        <div className="p-2"><img alt="" style={styles.hintImage} src={hintColor} /></div>
                                        <div className="p-2">
                                            <ColorPicker
                                                color={siteHeader && siteHeader.highlightColour}
                                                refresh={siteHeader && refreshColorPicker}
                                                handleColorChange={this.handleHeaderHighlightColorChange}
                                            />
                                        </div>
                                    </div>

                                </Col>
                            </Row>
                            <Row style={{ ...styles.RowParent, ...styles.RowParent2 }}>
                                <Col lg={{ size: 12 }}>
                                    <h4 style={styles.explainerHeader}>Branding</h4>
                                    <p>You can add your logo and title to the header, plus a favicon for the site.</p>

                                    <div style={styles.settingsParent} className="d-flex justify-content-start">
                                        <div className="p-2"><img alt="" style={styles.hintImage} src={hintHeader} /></div>
                                        <FileUploader
                                            url={siteHeader && siteHeader.logoUrl}
                                            file={siteHeader && siteHeader.logoFile}
                                            refresh={refreshHeaderLogo}
                                            updateFile={this.updateHeaderFile}
                                            uploadTitle="Page header logo"
                                            altTag="Logo"
                                            id="logoImage"
                                        />
                                    </div>
                                </Col>

                                <Col lg={{ size: 12 }}>
                                    <div style={styles.settingsParent} className="d-flex justify-content-start">
                                        <div className="p-2"><img alt="" style={styles.hintImage} src={hintBrowser} /></div>
                                        <div>
                                            <FileUploader
                                                url={siteHeader && siteHeader.faviconUrl}
                                                file={siteHeader && siteHeader.faviconFile}
                                                refresh={refreshFaviconImage}
                                                updateFile={this.updateFaviconFile}
                                                uploadTitle="Browser favicon icon"
                                                altTag="Favicon"
                                                id="faviconImage"
                                            />
                                        </div>
                                    </div>
                                </Col>
                                <Col style={styles.buttonParent} lg={{ size: 12 }}>
                                    <BtnRow
                                        buttonLabel={buttonTitle}
                                        cancelForm={this.cancelForm}
                                    />
                                </Col>
                            </Row>

                            <p style={{ ...styles.hint, ...styles.hintVisibile }} className="small">
                                Thank you, please refresh your screen to view saved changes
                            </p>

                        </div>

                    </AvForm>

                </Container>
            </div>
        );
    }
}

Config.contextType = StoreContext;

export default observer(Config);
