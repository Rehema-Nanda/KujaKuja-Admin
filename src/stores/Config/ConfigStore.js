import {
    flow, decorate, observable,
} from "mobx";
import Config from "./Config";
import ConfigApiService from "./ConfigApiService";
import AppConfig from "../../AppConfig";

class ConfigStore {
    siteHeader = {};
    isLoading = false;

    constructor(rootStore) {
        this.rootStore = rootStore;
        this.apiService = new ConfigApiService(AppConfig.API_URL);

        this.loadConfig();
    }

    loadConfig = flow(function* () { // eslint-disable-line func-names
        this.isLoading = true;
        try {
            const configJson = yield this.apiService.fetchConfig();
            const newConfig = new Config(
                this,
                configJson,
            );
            this.siteHeader = newConfig;
            this.isLoading = false;
            return this.siteHeader;
        }
        catch (error) {
            console.error(`ConfigStore failed to load config. ${error.message}`); // eslint-disable-line no-console
            this.isLoading = false;
        }
    });
}

decorate(ConfigStore, {
    siteHeader: observable,
    isLoading: observable,
});

export default ConfigStore;
