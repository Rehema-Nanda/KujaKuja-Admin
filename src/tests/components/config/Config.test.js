import React from "react";
import { mount } from "enzyme";
import ConfigComponent from "../../../components/config/Config";
import { StoreProvider } from "../../../StoreContext";
import ConfigStore from "../../../stores/Config/ConfigStore";
import Config from "../../../stores/Config/Config";


jest.mock("../../../stores/Config/ConfigStore");

const config = [
    new Config(
        null,
        {
            favicon_url: null,
            highlight_colour: "#FFF01",
            logo_url: null,
            title_text: "test title",
            logo_file: null,
            favicon_file: null,
        },
    ),
];
ConfigStore.mockImplementation(() => {
    return {
        siteHeader: config,
    };
});

describe("#config", () => {
    const wrapper = mount(<ConfigComponent />, { wrappingComponent: StoreProvider });

    it("should render correctly", () => {
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it("should toggle refreshHeaderLogo state when updateHeaderFile function is called", () => {
        wrapper.instance().simpleState.siteHeader = {};
        const updateHeaderFile = jest.spyOn(wrapper.instance(), "updateHeaderFile");
        const refreshHeaderLogoState = wrapper.state().refreshHeaderLogo;
        updateHeaderFile();
        expect(wrapper.state().refreshHeaderLogo).toBe(!refreshHeaderLogoState);
    });

    it("should toggle refreshFaviconImage state when updateFaviconFile function is called", () => {
        wrapper.instance().simpleState.siteHeader = {};
        const updateFaviconFile = jest.spyOn(wrapper.instance(), "updateFaviconFile");
        const refreshFaviconImage = wrapper.state().refreshFaviconImage;
        updateFaviconFile();
        expect(wrapper.state().refreshFaviconImage).toBe(!refreshFaviconImage);
    });

    it("should set highlight color when handleHeaderHighlightColorChange function is called", () => {
        wrapper.instance().simpleState.siteHeader = {};
        const color = {
            hex: "#abc",
        };
        const handleHeaderHighlightColorChange = jest.spyOn(wrapper.instance(), "handleHeaderHighlightColorChange");
        handleHeaderHighlightColorChange(color);
        expect(wrapper.instance().simpleState.siteHeader.highlightColour).toBe(color.hex);
    });

    it("should set title text when handleTitleTextChange function is called", () => {
        wrapper.instance().simpleState.siteHeader = {};
        const event = {
            target: {
                value: "testText",
            },
        };
        const handleTitleTextChange = jest.spyOn(wrapper.instance(), "handleTitleTextChange");
        handleTitleTextChange(event);
        expect(wrapper.instance().simpleState.siteHeader.titleText).toBe(event.target.value);
    });
});
