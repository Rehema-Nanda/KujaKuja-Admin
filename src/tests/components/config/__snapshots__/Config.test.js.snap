// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#config should render correctly 1`] = `
"<Config>
  <div className=\\"page\\">
    <div className=\\"title-background-parent\\">
      <div className=\\"title-background-shape\\" />
    </div>
    <Banner pageTitle=\\"custom header\\" buttonDisabled={false}>
      <Container tag=\\"div\\">
        <div className=\\"container\\">
          <Row tag=\\"div\\">
            <div className=\\"row\\">
              <Col sm={{...}} md={{...}} tag=\\"div\\" widths={{...}}>
                <div className=\\"col-sm-7 col-md-8\\">
                  <div className=\\"page-title\\">
                    <h1 className=\\"left\\">
                      Custom header
                    </h1>
                    <p>
                      Manage your
                       
                      custom header
                    </p>
                  </div>
                </div>
              </Col>
              <Col sm={{...}} md={{...}} tag=\\"div\\" widths={{...}}>
                <div className=\\"col-sm-5 col-md-4\\">
                  <div className=\\"page-title\\" />
                </div>
              </Col>
            </div>
          </Row>
        </div>
      </Container>
    </Banner>
    <Container tag=\\"div\\">
      <div className=\\"container\\">
        <AvForm encType=\\"multipart/form-data\\" onValidSubmit={[Function: bound handleFormSubmit]} model={{...}} tag={[Function: Form]} validationEvent={{...}} method=\\"get\\" onSubmit={[Function: onSubmit]} onKeyDown={[Function: onKeyDown]} onInvalidSubmit={[Function: onInvalidSubmit]}>
          <Form noValidate={true} action=\\"#\\" encType=\\"multipart/form-data\\" method=\\"get\\" onSubmit={[Function]} onKeyDown={[Function: onKeyDown]} className=\\"av-valid\\" tag=\\"form\\">
            <form noValidate={true} action=\\"#\\" encType=\\"multipart/form-data\\" method=\\"get\\" onSubmit={[Function]} onKeyDown={[Function: onKeyDown]} className=\\"av-valid\\">
              <div style={{...}}>
                <Row style={{...}} tag=\\"div\\">
                  <div style={{...}} className=\\"row\\">
                    <Col lg={{...}} tag=\\"div\\" widths={{...}}>
                      <div className=\\"col-lg-12\\">
                        <h4 style={{...}}>
                          Header Highlight Color
                        </h4>
                        <p>
                          This is the thin color band that runs across the very top of the page.
                        </p>
                        <div style={{...}} className=\\"d-flex justify-content-start\\">
                          <div className=\\"p-2\\">
                            <img alt=\\"\\" style={{...}} src=\\"hint-color-band-213px.png\\" />
                          </div>
                          <div className=\\"p-2\\">
                            <ColorPicker color={{...}} refresh={{...}} handleColorChange={[Function: bound handleHeaderHighlightColorChange]}>
                              <Row tag=\\"div\\">
                                <div className=\\"row\\">
                                  <Col style={{...}} sm=\\"12\\" md=\\"6\\" tag=\\"div\\" widths={{...}}>
                                    <div style={{...}} className=\\"col-sm-12 col-md-6\\">
                                      <div style={{...}} onClick={[Function]} role=\\"presentation\\">
                                        <div style={{...}} />
                                      </div>
                                    </div>
                                  </Col>
                                </div>
                              </Row>
                            </ColorPicker>
                          </div>
                        </div>
                      </div>
                    </Col>
                  </div>
                </Row>
                <Row style={{...}} tag=\\"div\\">
                  <div style={{...}} className=\\"row\\">
                    <Col lg={{...}} tag=\\"div\\" widths={{...}}>
                      <div className=\\"col-lg-12\\">
                        <h4 style={{...}}>
                          Branding
                        </h4>
                        <p>
                          You can add your logo and title to the header, plus a favicon for the site.
                        </p>
                        <div style={{...}} className=\\"d-flex justify-content-start\\">
                          <div className=\\"p-2\\">
                            <img alt=\\"\\" style={{...}} src=\\"hint-header-logo-220px.png\\" />
                          </div>
                          <FileUploader url={{...}} file={{...}} refresh={false} updateFile={[Function: bound updateHeaderFile]} uploadTitle=\\"Page header logo\\" altTag=\\"Logo\\" id=\\"logoImage\\">
                            <div style={{...}} className=\\"d-flex justify-content-start\\">
                              <div>
                                <div style={{...}}>
                                  <img alt=\\"Logo\\" id=\\"logoImage\\" style={{...}} src={{...}} />
                                  <div style={{...}}>
                                    <img alt=\\"Edit Icon\\" style={{...}} src={{...}} />
                                  </div>
                                </div>
                              </div>
                              <div style={{...}}>
                                <h3 style={{...}}>
                                  Page header logo
                                </h3>
                                <input name=\\"file\\" onChange={[Function]} type=\\"file\\" style={{...}} />
                              </div>
                            </div>
                          </FileUploader>
                        </div>
                      </div>
                    </Col>
                    <Col lg={{...}} tag=\\"div\\" widths={{...}}>
                      <div className=\\"col-lg-12\\">
                        <div style={{...}} className=\\"d-flex justify-content-start\\">
                          <div className=\\"p-2\\">
                            <img alt=\\"\\" style={{...}} src=\\"hint-browser-tab-214px.png\\" />
                          </div>
                          <div>
                            <FileUploader url={{...}} file={{...}} refresh={false} updateFile={[Function: bound updateFaviconFile]} uploadTitle=\\"Browser favicon icon\\" altTag=\\"Favicon\\" id=\\"faviconImage\\">
                              <div style={{...}} className=\\"d-flex justify-content-start\\">
                                <div>
                                  <div style={{...}}>
                                    <img alt=\\"Favicon\\" id=\\"faviconImage\\" style={{...}} src={{...}} />
                                    <div style={{...}}>
                                      <img alt=\\"Edit Icon\\" style={{...}} src={{...}} />
                                    </div>
                                  </div>
                                </div>
                                <div style={{...}}>
                                  <h3 style={{...}}>
                                    Browser favicon icon
                                  </h3>
                                  <input name=\\"file\\" onChange={[Function]} type=\\"file\\" style={{...}} />
                                </div>
                              </div>
                            </FileUploader>
                          </div>
                        </div>
                      </div>
                    </Col>
                    <Col style={{...}} lg={{...}} tag=\\"div\\" widths={{...}}>
                      <div style={{...}} className=\\"col-lg-12\\">
                        <BtnRow buttonLabel=\\"Update\\" cancelForm={[Function: bound cancelForm]}>
                          <Row className=\\"btn-row\\" tag=\\"div\\">
                            <div className=\\"btn-row row\\">
                              <Col tag=\\"div\\" widths={{...}}>
                                <div className=\\"col\\" />
                              </Col>
                              <Col tag=\\"div\\" widths={{...}}>
                                <div className=\\"col\\">
                                  <Button color=\\"primary\\" className=\\"primary cancel\\" onClick={[Function: bound cancelForm]} tag=\\"button\\">
                                    <button type=\\"button\\" onClick={[Function: bound onClick]} className=\\"primary cancel btn btn-primary\\" aria-label={{...}}>
                                      Back
                                    </button>
                                  </Button>
                                </div>
                              </Col>
                            </div>
                          </Row>
                        </BtnRow>
                      </div>
                    </Col>
                  </div>
                </Row>
                <p style={{...}} className=\\"small\\">
                  Thank you, please refresh your screen to view saved changes
                </p>
              </div>
            </form>
          </Form>
        </AvForm>
      </div>
    </Container>
  </div>
</Config>"
`;
